src/lsm/set_associative_cache.zig

```diff
@@ -0,0 +1,423 @@
+        };
+
+        pub fn get_or_put(self: *Self, key: Key) GetOrPutResult {
+            return self.get_or_put_preserve_locked(
+                void,
+                struct {
+                    inline fn locked(_: void, _: *const Value) bool {
+                        return false;
+                    }
+                }.locked,
+                {},
+                key,
+            );
+        }
+
+        /// Get a pointer to an existing value with the given key if the key is in the cache.
+        /// If the key is not in the cache, add it to the cache, evicting older entries if needed.
+        /// Never evicts keys for which locked() returns true.
+        /// The caller must guarantee that locked() returns true for less than layout.ways keys.
+        pub fn get_or_put_preserve_locked(
+            self: *Self,
+            comptime Context: type,
+            comptime locked: fn (Context, *align(value_alignment) const Value) callconv(.Inline) bool,
+            context: Context,
+            key: Key,
+        ) GetOrPutResult {
+            const set = self.associate(key);
+
+            if (self.search(set, key)) |existing| {
+                assert(!locked(context, existing));
+                return .{
+                    .value_ptr = existing,
+                    .found_existing = true,
+                };
+            }
+
+            const clock_index = @divExact(set.offset, layout.ways);
+
+            const max_iterations = layout.ways * math.maxInt(Count);
+            var safety_counter: usize = 0;
+
+            var way = Clocks.get(self.clocks, clock_index);
+            while (safety_counter < max_iterations + 1) : ({
+                way +%= 1;
+                safety_counter += 1;
+            }) {
+                // We pass a value pointer to the callback here so that a cache miss
+                // can be avoided if the caller is able to determine if the value is
+                // locked by comparing pointers directly.
+                if (locked(context, @alignCast(value_alignment, &set.values[way]))) continue;
+
+                const count = Counts.get(self.counts, set.offset + way);
+
+                // Free way found
+                if (count == 0) break;
+
+                Counts.set(self.counts, set.offset + way, count - 1);
+            }
+            assert(safety_counter < max_iterations);
+
+            Clocks.set(self.clocks, clock_index, way +% 1);
+
+            set.tags[way] = set.tag;
+
+            Counts.set(self.counts, set.offset + way, 1);
+
+            return .{
+                .value_ptr = @alignCast(value_alignment, &set.values[way]),
+                .found_existing = false,
+            };
+        }
+
+        fn search(self: *Self, set: Set, key: Key) ?*align(value_alignment) Value {
+            for (set.tags) |tag, way| {
+                // TODO self.counts.get()
+                const count = Counts.get(self.counts, set.offset + way);
```

Comment:
I missed this yesterday, and for the first draft last year didn't think about it beyond branching on the tag first—but the `const count` assignment must be moved to be within the `if (tag == set.tag)` branch, with a separate load of the count to evaluate the `count > 0` branch only if necessary.

The reason is that the tag contains enough entropy to avoid the second branch most of the time, so we don't even need to lookup the relevant count.

Reply:
There's nothing stopping llvm from reordering the `Counts.get()` call after the `tag == set.tag` comparison with the current code, and my intuition says it would do so. That said, I'm happy to change the code and make this explicit, good thought!

Reply:
bbda2adfba88543a8af728dc05a93806f9af6464